### PR TITLE
Sort, tree view and optimisation

### PR DESCRIPTION
**Pull request by TheLortex**
image

## Sort abstraction

I have opted for a straightforward solution for now, which is to sort the table once per frame. It might be possible to maintain the sorted invariant in the table but it's complicated and I don't know yet if there's a benefit. It's complicated in particular for the case of the tree view, which requires a topological sort. The sorting mechanism is abstracted as with the following type:
```ocaml
type sort =
  | V : {
      prepare : ((Task.t -> unit) -> unit) -> 'a;
      compare : 'a -> Task.t -> Task.t -> int;
    }
      -> sort
```
`prepare` enables to inject an implementation of the topological sort for the tree view and pre-compute the order. 
`compare` is a row comparison function that can make use of the prepared value.

Fixes https://github.com/tarides/meio/issues/23

## Tree view

Implemented by looking at the current fiber ID when a task is created, so that there is a parent_id -> id mapping. Then it's sufficient to explore the tree starting from ID -1. 

Fixes https://github.com/tarides/meio/issues/28
(but doesn't contain switches yet)

## Optimisation

`update_active` is not very efficient as it goes through the fiber list to find which one to update. In `Task_table.t` I'm adding a `by_id` hashtable to enable a fast lookup of rows by ID. We need to be cautious to keep this table synchronized when rows move around (for example during sort).
With this patch the interface doesn't suffer slowdowns increasing with the number of fibers.

### CODE REVIEW

**Comment by patricoferris** [comment]
Nice! 

On the sorting, I totally agree. I'm not 100% sure how useful it is. Perhaps as a "please sort the fibres right now by busy time" but not maintaining that invariant, they jump around too much which makes it hard to follow.

> Implemented by looking at the current fiber ID when a task is created, so that there is a parent_id -> id mapping. 

Have you run into any troubles with the ID not being domain-safe in Eio yet?

https://github.com/TheLortex/eio/blob/e6f3e01e07d340f3aef9252ad2b529c60ad2fcad/lib_eio/core/ctf.ml#L31-L35

I don't know if it is better to make that slow and atomic or wait until it becomes a problem.

> update_active is not very efficient as it goes through the fiber list to find which one to update

This seems very sensible to optimise thank you (apologies for the slow code!). I wonder if we should add a comment to `task_table.ml` about keeping things in sync.

**Comment by TheLortex** [comment]
> Have you run into any troubles with the ID not being domain-safe in Eio yet?

I haven't tested so much with multicore programs, but I expect it to break if IDs are not domain safe.  
A strategy for generating IDs is to keep N bits for the domain ID and have domain local counters. For that we need to know the maximum number of domains that can be spawned. 

> I wonder if we should add a comment to task_table.ml about keeping things in sync.

Good idea.
